File: state.cpp
#include <random>
#include <iostream>
#include <array>
#include <list>
#include <map>
#include <unordered_set>
#include <algorithm>
using namespace std;

struct node { 
    array<int, 24> board; 
    map<int, int> bar;
    list<int> towers;
    int depth;
};

struct node_comparator {
    bool operator()(const node &lhs, const node &rhs) const {
        if (lhs.board == rhs.board && lhs.bar == rhs.bar && lhs.depth == rhs.depth){
            return true;
        } else {
            return false;
        }
    }
};

struct node_hasher {
    int operator()(const node &n) const {
        string res = "";
        int size = n.board.size();
        for (auto const& i : n.board){
            res += to_string(i);
        }
        for (auto const& kv : n.bar){
            res += to_string(kv.second);
        }
        return hash<string>()(res);
    }
};

node create_node(array<int, 24> board, map<int, int> bar, list<int> towers, int depth){
    node n;
    n.board = board;
    n.bar = bar;
    n.towers = towers;
    n.depth = depth;
    return n;
}

auto roll_dice() {
    random_device rd; 
    mt19937 gen(rd()); 
    uniform_int_distribution<> dis(1, 6);

    int r1 = dis(gen);
    int r2 = dis(gen);
    if (r1 == r2){
        return vector<int> {r1, r1, r1, r1};
    } else {
        return vector<int> {r1, r2};
    }
}

bool is_valid_move(int t, int r, array<int, 24> board, list<int> towers) {
    int end = t + r;
    if (end <= 23){
        if (board[end] >= -1){
            return true;
        } else {
            return false;
        }
    } else {
        int t_min = *min_element(towers.begin(), towers.end());
        if (t_min >= 18 && (r == 24 - t || t == t_min)){
            return true;
        } else {
            return false;
        }
    }
}

void add_piece(array<int, 24>& board, map<int, int>& bar, int p, list<int>& towers){
    if (bar[1] > 0){
        bar[1] -= 1;
    }
    if (board[p] >= 0){
        board[p] += 1;
    }
    if (board[p] == -1){
        board[p] = 1;
        bar[-1] += 1;
    }
    if (board[p] == 1){
        towers.push_back(p);
    }
}

struct State{
    vector<int> roll = roll_dice();
    array<int, 24> board = {2, 0, 0, 0, 0, -5, 0, -3, 0, 0, 0, 5, -5, 0, 0, 0, 3, 0, 5, 0, 0, 0, 0, -2};
    map<int, int> bar = {{1, 0}, {-1, 0}};
    int max_depth = roll.size();
    list<int> towers;
    unordered_set<node, node_hasher, node_comparator> leaves;
    int winner = 0;
    int sign = 1;
    
    void extend(node n){
        array<int, 24> board_;
        map<int, int> bar_;
        list<int> towers_;
        node n_new;

        if (winner){
           return;
        }

        if (n.depth == max_depth){
            leaves.insert(n);
            return;
        }

        int r = roll[n.depth];
        if (n.bar[1] != 0){
            if (n.board[r-1] >= -1){
                tie(board_, bar_, towers_) = make_tuple(n.board, n.bar, n.towers);
                add_piece(board_, bar_, r-1, towers_);
                n_new = create_node(board_, bar_, towers_, n.depth+1);
                extend(n_new);
            } else {
                if (max_depth == 2){
                    n_new = n;
                    n_new.depth += 1;
                    extend(n_new);
                } else {
                    leaves.insert(n);
                }
            }
        } else {
            bool is_leaf = true;
            for (auto const& t : n.towers){
                if (is_valid_move(t, r, n.board, n.towers)){
                    is_leaf = false;
                    tie(board_, bar_, towers_) = make_tuple(n.board, n.bar, n.towers);
                    board_[t] -= 1;
                    if (board_[t] == 0){
                        towers_.remove(t);
                    }
                    int end = t + r;
                    if (end <= 23){
                        add_piece(board_, bar_, end, towers_);
                    }
                    if (towers_.size() == 0){
                        winner = sign;
                        leaves.insert(n);
                        return;
                    }
                    n_new = create_node(board_, bar_, towers_, n.depth+1);
                    extend(n_new);
                }
            }
            if (is_leaf){
                leaves.insert(n);
            }
        }
    }

    auto get_moves(){
		list<int> towers;
        for (int i = 0; i < board.size(); ++i){
            if (board[i] > 0){
                towers.push_back(i);
            }
        }
        
        for (const auto& t: towers){
            cout << t << " ";
        }
        
        int depth = 0;  
        node root = create_node(board, bar, towers, depth);
        extend(root);
        if (roll.size() == 2){
            roll = {roll[1], roll[0]};
            extend(root);
        }
        
        int d_max = 0;
        for (const auto& n: leaves) {
            if (n.depth > d_max){
                d_max = n.depth;
            }
        }
        
        list<pair<array<int, 24>, map<int, int>>> res;
        for (const auto& n: leaves) {
            if (n.depth == d_max){
                res.push_back({n.board, n.bar});
            }
        }
        return res;
    }
    
    void update(pair<array<int, 24>, map<int, int>> move){
        board = move.first;
        bar = move.second;
        for (int i = 0; i < board.size(); ++i){
            board[i] *= -1;
        }
        reverse(board.begin(), board.end());
        
        int q = bar[-1];
        bar[-1] = bar[1];
        bar[1] = q;
        
        roll = roll_dice();
        max_depth = roll.size();
        sign *= -1;
    }
	
	void update_randomly(){
        list<pair<array<int, 24>, map<int, int>>> moves = get_moves();

        vector<pair<array<int, 24>, map<int, int>>> out;
        random_device rd; 
        mt19937 g(rd()); 
        sample(moves.begin(), moves.end(), back_inserter(out), 1, g);
        
        board = out[0].first;
        bar = out[0].second;
        for (int i = 0; i < board.size(); ++i){
            board[i] *= -1;
        }
        reverse(board.begin(), board.end());
        
        int q = bar[-1];
        bar[-1] = bar[1];
        bar[1] = q;
        
        roll = roll_dice();
        max_depth = roll.size();
        sign *= -1;
    }
};

void pit(){
	while (true){
		s.update_randomly();
		if (s.winner){
			break;
		}
	}
}

File: state.py
import random
from collections import namedtuple

Node = namedtuple('Node', ['board', 'bar', 'towers', 'depth'])


def roll_dice():
    r1, r2 = random.randint(1, 6), random.randint(1, 6)
    return (r1, r2) if r1 != r2 else (r1,) * 4


def is_valid_move(t, r, board, towers):
    end = t + r
    if end <= 23:
        if board[end] >= -1:  # <= 1 фигур противника
            return True
        return False
    else:
        t_min = min(towers)
        if t_min >= 18 and (r == 24 - t or t == t_min):
            return True
        return False


def add_piece(board, bar, p, towers):
    assert p <= 23, 'invalid position'
    assert board[p] >= -1, 'position is occupied'
    if bar[1] > 0:
        bar[1] -= 1  # убрать шашку с bar
    if board[p] >= 0:  # не противник
        board[p] += 1  # добавить шашку
    if board[p] == -1:  # одна шашка противника
        board[p] = 1  # поставить одну шашку
        bar[-1] += 1  # добавить одну шашку противника в bar
    if board[p] == 1:  # если p не была в towers
        towers.append(p)


class State:
    """
    Класс состояния игры. Характеризуется:
    - выпавшими кубиками,
    - расположением фигур на доске,
    - съеденными фигурами
    """

    def __init__(self):
        self.board = [2, 0, 0, 0, 0, -5, 0, -3, 0, 0, 0, 5, -5, 0, 0, 0, 3, 0, 5, 0, 0, 0, 0, -2]
        self.bar = {-1: 0, 1: 0}
        self.roll = roll_dice()
        while len(self.roll) == 4:
            self.roll = roll_dice()
        self.towers = None
        self.winner = None
        self.sign = 1  # нужен только для определения победителя

    def reverse_state(self):
        self.board = [-x for x in self.board[::-1]]
        self.bar = {-k: v for k, v in self.bar.items()}

    def get_moves_roll(self, roll):
        leaves = []
        max_depth = len(roll)

        def extend(node):
            if self.winner:
                return

            leaf_data = tuple(node.board), tuple(node.bar.items()), None, node.depth

            if node.depth == max_depth:
                leaves.append(Node(*leaf_data))
                return

            r = roll[node.depth]
            if node.bar[1]:
                board_, bar_, towers_ = node.board.copy(), node.bar.copy(), node.towers.copy()
                if node.board[r - 1] >= -1:
                    add_piece(board_, bar_, r - 1, towers_)
                    extend(Node(board_, bar_, towers_, node.depth + 1))
                else:
                    if max_depth == 2:
                        extend(Node(board_, bar_, towers_, node.depth + 1))
                    else:
                        leaves.append(Node(*leaf_data))
            else:
                is_leaf = True
                for t in node.towers:
                    if is_valid_move(t, r, node.board, node.towers):
                        is_leaf = False
                        board_, bar_, towers_ = node.board.copy(), node.bar.copy(), node.towers.copy()
                        board_[t] -= 1
                        if board_[t] == 0:
                            towers_.remove(t)
                        end = t + r
                        if end <= 23:
                            add_piece(board_, bar_, end, towers_)
                        if len(towers_) == 0:
                            self.winner = self.sign
                            leaves.append(Node(*leaf_data))
                            return
                        extend(Node(board_, bar_, towers_, node.depth + 1))
                if is_leaf:
                    leaves.append(Node(*leaf_data))

        root = Node(self.board, self.bar, self.towers, 0)
        extend(root)
        if self.winner:
            return leaves[-1]
        return leaves if leaves else [Node(tuple(self.board), tuple(self.bar.items()), None, 0)]

    def get_moves(self):
        self.towers = [i for i, x in enumerate(self.board) if x > 0]
        leaves = self.get_moves_roll(self.roll)
        if self.winner:
            return [(list(leaves.board), dict(leaves.bar))]
        if len(self.roll) == 2:
            leaves_ = self.get_moves_roll(self.roll[::-1])
            if self.winner:
                return [(list(leaves_.board), dict(leaves_.bar))]
            leaves += leaves_
        leaves = set(leaves)
        max_depth = max(leaf.depth for leaf in leaves)
        return [(list(leaf.board), dict(leaf.bar)) for leaf in leaves if leaf.depth == max_depth]

    def clone(self):
        s = State()
        for attr, value in vars(self).items():
            setattr(s, attr, value)
        return s

    def update(self, move):
        self.board, self.bar = move  # обновление доски
        self.sign *= -1  # обновление знака текущего игрока
        self.reverse_state()  # разворот доски к новому игрока
        self.roll = roll_dice()  # обновление кубиков


if __name__ == '__main__':
    """
    1 1 9
    1 2 15
    1 3 16
    1 4 14
    1 5 8
    1 6 10
    2 2 13
    2 3 17
    2 4 18
    2 5 8
    2 6 14
    3 3 13
    3 4 17
    3 5 9
    3 6 14
    4 4 12
    4 5 9
    4 6 14
    5 5 3
    5 6 7
    6 6 6
    """
    s = State()
    for i in range(1, 7):
        for j in range(i, 7):
            s.roll = i, j
            print(i, j, len(s.get_moves()))
